pkg/instance/service.go

```diff
@@ -869,3 +812,106 @@ func (s service) destroy(instance *model.Instance) error {
 	}
 	return nil
 }
+
+type updateMessage struct {
+	EventType    watch.EventType `json:"eventType"`
+	DeploymentId uint            `json:"deploymentId"`
+	InstanceId   uint            `json:"instanceId"`
+	Status       Status          `json:"status"`
+}
+
+func (s service) ListenForClusterUpdates() {
+	fakeAdministrator := &model.User{
+		Groups: []model.Group{
+			{
+				Name: model.AdministratorGroupName,
+			},
+		},
+	}
+
+	groups, err := s.groupService.FindAll(fakeAdministrator, true)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	for _, group := range groups {
+		g := group
+		go func() {
+			ks, err := NewKubernetesService(g.ClusterConfiguration)
+			if err != nil {
+				log.Printf("unable to get create new kubernetes service: %v", err)
+			}
+
+			for {
+				events, err := ks.watch(g.Name)
+				if err != nil {
+					log.Printf("error watching %q: %v", g.Name, err)
+				}
+
+				for e := range events {
+					s.handleEvent(e)
+				}
+			}
+		}()
+	}
+}
+
+func (s service) handleEvent(e watch.Event) {
+	pod, ok := e.Object.(*v1.Pod)
+	if !ok {
+		// TODO: Log entire e.Object... Entire e?
+		log.Println("unexpected type")
+	}
+
+	status, err := PodStatus(pod)
+	if err != nil {
+		log.Printf("unable to get pod status: %v", err)
+	}
+
+	deploymentIdStr, ok := pod.Labels["im-deployment-id"]
+	if !ok {
+		log.Println("im-deployment-id label not found")
+	}
+	deploymentId, err := strconv.Atoi(deploymentIdStr)
+	if err != nil {
+		log.Printf("can't convert deployment id to integer: %v", err)
+	}
+
+	instanceIdStr, ok := pod.Labels["im-instance-id"]
+	if !ok {
+		log.Println("im-instance-id label not found")
+	}
+
+	instanceId, err := strconv.Atoi(instanceIdStr)
+	if err != nil {
+		log.Printf("can't convert instance id to integer: %v", err)
+	}
+
+	err = s.UpdateInstanceStatus(uint(instanceId), status)
+	if err != nil {
+		log.Printf("unable to update instance status: %v", err)
+	}
+
+	message := updateMessage{
+		// TODO: We probably shouldn't return the raw e.Type
+		EventType:    e.Type,
+		DeploymentId: uint(deploymentId),
+		InstanceId:   uint(instanceId),
+		Status:       status,
+	}
+
+	jsonMessage, err := json.Marshal(message)
+	if err != nil {
+		log.Printf("failed to marshal message: %v", err)
+	}
+
+	subscribers := s.broker.Subscribers()
+	for _, subscriber := range subscribers {
+		if subscriber.IsMemberOf(pod.Namespace) {
+			s.broker.Send(subscriber.ID, event.Event{
```

Comment:
I might not have the full picture yet of the flow but I think: any subscriber that is not able to receive will block (when the buffer of 1 is full) any other subscriber that would be able to receive the event.